kernel/kernel-api/src/main/java/io/delta/kernel/internal/snapshot/SnapshotManager.java

```diff
@@ -306,48 +318,93 @@ protected final Optional<List<FileStatus>> listDeltaAndCheckpointFiles(
+                        final boolean versionWithinRange = versionToLoad
+                                .map(v -> FileNames
+                                        .getFileVersion(new Path(fileStatus.getPath())) <= v)
+                                .orElse(true);
+                        if (!versionWithinRange) {
+                            // If we haven't taken any files yet and the first file we see is
+                            // greater than the versionToLoad then the versionToLoad is not
+                            // reconstructable from the existing logs
+                            if (output.isEmpty()) {
+                                long earliestVersion = DeltaHistoryManager
+                                        .getEarliestRecreatableCommit(engine, logPath);
+                                throw DeltaErrors.versionBeforeFirstAvailableCommit(
+                                        tablePath.toString(), versionToLoad.get(), earliestVersion);
+                            }
+                            break;
+                        }
 
-                    // Take files until the version we want to load
-                    final boolean versionWithinRange = versionToLoad
-                        .map(v -> FileNames.getFileVersion(new Path(fileStatus.getPath())) <= v)
-                        .orElse(true);
-
-                    if (!versionWithinRange) {
-                        // If we haven't taken any files yet and the first file we see is greater
-                        // than the versionToLoad then the versionToLoad is not reconstructable
-                        // from the existing logs
-                        if (output.isEmpty()) {
-                            long earliestVersion = DeltaHistoryManager.getEarliestRecreatableCommit(
-                                engine, logPath);
-                            throw DeltaErrors.versionBeforeFirstAvailableCommit(
-                                tablePath.toString(), versionToLoad.get(), earliestVersion);
+                        // Ideally listFromOrNone should return lexiographically sorted files amd so
+                        // maxDeltaVersionSeen should be equal to fileVersion.
+                        // But we are being defensive here and taking max of all the fileVersions
+                        // seen.
+                        if (FileNames.isCommitFile(fileName)) {
+                            maxDeltaVersionSeen.set(Math.max(
+                                    maxDeltaVersionSeen.get(),
+                                    FileNames.deltaVersion(fileStatus.getPath())));
                         }
-                        break;
+
+                        output.add(fileStatus);
                     }
 
-                    output.add(fileStatus);
-                }
+                    return output;
+                });
 
-                return output;
-            });
+        if (!commitCoordinatorClientHandlerOpt.isPresent()) {
```

Comment:
can this be `commitCoordinatorClientHandlerOpt.isEmpty() ?

Reply:
It seems that `Optional#isEmpty()` is a Java 11 method, which is a shortcut for Java 8's `!Optional#isPresent()`?